plugins/vbe/fit.c

```diff
@@ -0,0 +1,346 @@
+	if (subnode < 0)
+		return -FITE_NOT_FOUND;
+
+	return subnode;
+}
+
+const char *
+fit_cfg_name(struct fit_info *fit, int cfg)
+{
+	return fdt_get_name(fit->blob, cfg, NULL);
+}
+
+const char *
+fit_cfg_compat_item(struct fit_info *fit, int cfg, int index)
+{
+	return fdt_stringlist_get(fit->blob, cfg, FIT_PROP_COMPATIBLE, index, NULL);
+}
+
+int
+fit_cfg_img_count(struct fit_info *fit, int cfg, const char *prop_name)
+{
+	int count;
+
+	count = fdt_stringlist_count(fit->blob, cfg, prop_name);
+	if (count < 0)
+		return -FITE_NOT_FOUND;
+
+	return count;
+}
+
+int
+fit_cfg_img(struct fit_info *fit, int cfg, const char *prop_name, int index)
+{
+	const char *name;
+	int images, image;
+
+	name = fdt_stringlist_get(fit->blob, cfg, prop_name, index, NULL);
+	if (!name)
+		return -FITE_NOT_FOUND;
+
+	images = fdt_path_offset(fit->blob, FIT_IMAGE_PATH);
+	if (images < 0)
+		return -FITE_NO_IMAGES_NODE;
+
+	image = fdt_subnode_offset(fit->blob, images, name);
+	if (image < 0)
+		return -FITE_MISSING_IMAGE;
+
+	return image;
+}
+
+const char *
+fit_cfg_version(struct fit_info *fit, int cfg)
+{
+	return fdt_getprop(fit->blob, cfg, FIT_VERSION, NULL);
+}
+
+const char *
+fit_img_name(struct fit_info *fit, int img)
+{
+	return fdt_get_name(fit->blob, img, NULL);
+}
+
+static enum fit_algo_t
+fit_get_algo(struct fit_info *fit, int node)
+{
+	const char *algo;
+	int i;
+
+	algo = fdt_getprop(fit->blob, node, FIT_PROP_ALGO, NULL);
+	if (!algo)
+		return -FITE_MISSING_ALGO;
+
+	for (i = 0; i < FIT_ALGO_COUNT; i++) {
+		if (!strcmp(fit_algo[i], algo))
+			return i;
+	}
+
+	return -FITE_UNKNOWN_ALGO;
+}
+
+int
+fit_check_hash(struct fit_info *fit, int node, const char *data, int size)
+{
+	int algo;
+	const char *value;
+	int val_size;
+
+	value = fdt_getprop(fit->blob, node, FIT_PROP_VALUE, &val_size);
+	if (!value)
+		return -FITE_MISSING_VALUE;
+
+	/* Only check the algo after we have found a value */
+	algo = (int)fit_get_algo(fit, node);
+	if (algo < 0)
+		return algo;
+
+	switch (algo) {
+	case FIT_ALGO_CRC32: {
+		unsigned long actual;
+		unsigned long expect;
+
+		if (val_size != 4)
+			return -FITE_INVALID_HASH_SIZE;
+		expect = fdt32_to_cpu(*(fdt32_t *)value);
+		actual = crc32(0, (unsigned char *)data, size);
+		if (expect != actual)
+			return -FITE_HASH_MISMATCH;
+		break;
+	}
+	default:
+		return -FITE_UNKNOWN_ALGO;
+	}
+
+	return 0;
+}
+
+int
+fit_check_hashes(struct fit_info *fit, int img, const char *data, int size)
+{
+	int node;
+	int ret;
+
+	for (node = fdt_first_subnode(fit->blob, img); node > 0;
+	     node = fdt_next_subnode(fit->blob, node)) {
+		if (!strncmp("hash", fdt_get_name(fit->blob, node, NULL), 4)) {
+			ret = fit_check_hash(fit, node, data, size);
+
+			/* If the value is missing, we don't check it */
+			if (ret && ret != -FITE_MISSING_VALUE)
+				return ret;
+		}
+	}
+
+	return 0;
+}
+
+const char *
+fit_img_data(struct fit_info *fit, int img, int *sizep)
+{
+	const char *data;
+	int offset, size;
+	int ret;
```

Comment:
I agree with the static analysis here; change the scope of this variable

Reply:
done